src/bootstrap/doc.rs

```diff
@@ -551,6 +551,53 @@ fn doc_std(
     extra_args: &[&OsStr],
     requested_crates: &[String],
 ) {
+    // `cargo` uses the same directory for both JSON docs and HTML docs.
+    // This could lead to cross-contamination when copying files into the specified `out` directory.
+    // For example:
+    // ```bash
+    // x doc std
+    // x doc std --json
+    // ```
+    // could lead to HTML docs being copied into the JSON docs output directory.
+    // To avoid this issue, we copy generated docs instead of whole directory by
+    // checking doc format and generated files.
+    fn cp_docs_by_doc_format(
+        format: &DocumentationFormat,
+        builder: &Builder<'_>,
+        src: &Path,
+        dst: &Path,
+    ) {
+        if builder.config.dry_run {
+            return;
+        }
+
+        for f in builder.read_dir(src) {
+            let path = f.path();
+            let name = path.file_name().unwrap();
+            let dst = dst.join(name);
+
+            if t!(f.file_type()).is_dir() && format == &DocumentationFormat::HTML {
+                t!(fs::create_dir_all(&dst));
+                cp_docs_by_doc_format(format, builder, &path, &dst);
+            } else {
+                let _ = fs::remove_file(&dst);
+                let extension = path.extension().and_then(OsStr::to_str);
+
+                match format {
+                    DocumentationFormat::HTML if extension != Some("json") => {
+                        builder.copy(&path, &dst)
+                    }
+                    DocumentationFormat::JSON
+                        if extension == Some("json") || name.to_str() == Some(".stamp") =>
+                    {
+                        builder.copy(&path, &dst)
+                    }
```

Comment:
I am not sure if this conditions will be enough for the purpose. What do you think ? @jyn514